pkg/apis/tensorflow/v1alpha1/types.go

```diff
@@ -0,0 +1,195 @@
+type TfReplicaType string
+
+const (
+	MASTER TfReplicaType = "MASTER"
+	PS     TfReplicaType = "PS"
+	WORKER TfReplicaType = "WORKER"
+)
+
+// ContainerName is an enum for expected containers.
+type ContainerName string
+
+const (
+	TENSORFLOW     ContainerName = "tensorflow"
+	DefaultTFImage string        = "tensorflow/tensorflow:1.3.0"
+)
+
+// TODO(jlewi): We probably want to add a name field. This would allow us to have more than 1 type of each worker.
+// This might be useful if you wanted to have a separate set of workers to do eval.
+type TfReplicaSpec struct {
+	// Replicas is the number of desired replicas.
+	// This is a pointer to distinguish between explicit zero and unspecified.
+	// Defaults to 1.
+	// More info: http://kubernetes.io/docs/user-guide/replication-controller#what-is-a-replication-controller
+	// +optional
+	Replicas *int32              `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"`
+	Template *v1.PodTemplateSpec `json:"template,omitempty" protobuf:"bytes,3,opt,name=template"`
+	// TfPort is the port to use for TF services.
+	TfPort *int32 `json:"tfPort,omitempty" protobuf:"varint,1,opt,name=tfPort"`
+	TfReplicaType `json:"tfReplicaType"`
+	// IsDefaultPS denotes if the parameter server should use the default grpc_tensorflow_server
+	IsDefaultPS bool
+}
+
+type TensorBoardSpec struct {
+	//Location of TensorFlow event files
+	LogDir       string           `json:"logDir"`
+	Volumes      []v1.Volume      `json:"volumes"`
+	VolumeMounts []v1.VolumeMount `json:"volumeMounts"`
+	ServiceType  v1.ServiceType   `json:"serviceType"`
+}
+
+type TfJobPhase string
+
+const (
+	TfJobPhaseNone     TfJobPhase = ""
+	TfJobPhaseCreating TfJobPhase = "Creating"
+	TfJobPhaseRunning  TfJobPhase = "Running"
+	TfJobPhaseCleanUp  TfJobPhase = "CleanUp"
+	TfJobPhaseFailed   TfJobPhase = "Failed"
+	TfJobPhaseDone     TfJobPhase = "Done"
+)
+
+type State string
+
+const (
+	StateUnknown   State = "Unknown"
+	StateRunning   State = "Running"
+	StateSucceeded State = "Succeeded"
+	StateFailed    State = "Failed"
+)
+
+type TfJobStatus struct {
+	// Phase is the TfJob running phase
+	Phase  TfJobPhase `json:"phase"`
+	Reason string     `json:"reason"`
+
+	// State indicates the state of the job.
+	State State `json:"state"`
+
+	// ReplicaStatuses specifies the status of each TF replica.
+	ReplicaStatuses []*TfReplicaStatus `json:"replicaStatuses"`
+}
+
+type ReplicaState string
+
+const (
+	ReplicaStateUnknown   ReplicaState = "Unknown"
+	ReplicaStateStarting  ReplicaState = "Starting"
+	ReplicaStateRunning   ReplicaState = "Running"
+	ReplicaStateFailed    ReplicaState = "Failed"
+	ReplicaStateSucceeded ReplicaState = "Succeeded"
+)
+
+type TfReplicaStatus struct {
+	TfReplicaType `json:"tf_replica_type"`
+
+	// State is the overall state of the replica
+	State ReplicaState `json:"state"`
+
+	// ReplicasStates provides the number of replicas in each status.
+	ReplicasStates map[ReplicaState]int
```

Comment:
Nit: why there is so many `states`, State/ReplicaState/TfReplicaStatus

Reply:
Can you open up an issue for this?

Reply:
Sure

Reply:
FYI https://github.com/tensorflow/k8s/issues/249